Foundation/URLSession/NativeProtocol.swift

```diff
@@ -0,0 +1,754 @@
+        switch self {
+        case .partial(let header): return header
+        case .complete: return _NativeProtocol._ResponseHeaderLines()
+        }
+    }
+}
+
+extension _NativeProtocol._ResponseHeaderLines {
+    /// Returns a copy of the lines with the new line appended to it.
+    func byAppending(headerLine line: String) -> _NativeProtocol._ResponseHeaderLines {
+        var l = self.lines
+        l.append(line)
+        return _NativeProtocol._ResponseHeaderLines(headerLines: l)
+    }
+}
+
+extension _NativeProtocol._TransferState {
+    var isHeaderComplete: Bool {
+        return response != nil
+    }
+    func byAppending(bodyData buffer: Data) -> _NativeProtocol._TransferState {
+        switch bodyDataDrain {
+        case .inMemory(let bodyData):
+            let data: NSMutableData = bodyData ?? NSMutableData()
+            data.append(buffer)
+            let drain = _NativeProtocol._DataDrain.inMemory(data)
+            return _NativeProtocol._TransferState(url: url, parsedResponseHeader: parsedResponseHeader, response: response, requestBodySource: requestBodySource, bodyDataDrain: drain)
+        case .toFile(_, let fileHandle):
+            //TODO: Create / open the file for writing
+            // Append to the file
+            _ = fileHandle!.seekToEndOfFile()
+            fileHandle!.write(buffer)
+            return self
+        case .ignore:
+            return self
+        }
+    }
+    /// Sets the given body source on the transfer state.
+    ///
+    /// This can be used to either set the initial body source, or to reset it
+    /// e.g. when restarting a transfer.
+    func bySetting(bodySource newSource: _BodySource) -> _NativeProtocol._TransferState {
+        return _NativeProtocol._TransferState(url: url, parsedResponseHeader: parsedResponseHeader, response: response, requestBodySource: newSource, bodyDataDrain: bodyDataDrain)
+    }
+}
+
+extension _FTPURLProtocol._TransferState {
+   
+    enum FTPHeaderCode: String {
+        case transferCompleted = "226"
+        case openDataConnection = "150"
+        case fileStatus = "213"
+        case syntaxError = "5" // 500 series FTP Syntax errors
+        case errorOccurred = "4" // 400 Series FTP transfer errors
+    }
+ 
+    /// Appends a header line
+    ///
+    /// Will set the complete response once the header is complete, i.e. the
+    /// return value's `isHeaderComplete` will then by `true`.
+    ///
+    /// - Throws: When a parsing error occurs
+    func byAppendingFTP(headerLine data: Data, expectedContentLength: Int64) throws -> _NativeProtocol._TransferState {
+        let line = String(data: data, encoding: String.Encoding.utf8)
+        if (line?.starts(with: FTPHeaderCode.transferCompleted.rawValue))! {
+            return self
+        }
+        
+        func  isCompleteHeader(_ headerLine: String) ->Bool {
+            return  headerLine.starts(with: FTPHeaderCode.openDataConnection.rawValue)
```

Comment:
Double space after `return` keyword here - worth checking and replacing `return_space__space_` with `return_space_`